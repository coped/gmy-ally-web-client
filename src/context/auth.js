import React, { createContext, useContext, useState } from "react";

const AuthContext = createContext();

function AuthProvider(props) {
  const existingAuth = JSON.parse(localStorage.getItem("auth"));
  const [auth, setAuth] = useState(existingAuth);

  function setAuthContext(data) {
    setAuth(data);
    if (data === null) {
      localStorage.clear("auth");
    } else {
      localStorage.setItem("auth", JSON.stringify(data));
    }
  }
  return <AuthContext.Provider value={{ auth, setAuthContext }} {...props} />;
}

const useAuth = () => useContext(AuthContext);

export { AuthProvider, useAuth };
