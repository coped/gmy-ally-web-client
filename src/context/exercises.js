import React, { createContext, useContext, useState } from "react";

const ExercisesContext = createContext();

function ExercisesProvider(props) {
  const [allExercises, setAllExercises] = useState(null);
  const muscleGroups = getMuscleGroups();
  const categorizedExercises = getCategorizedExercises();

  // Get all unique muscle groups from exercises
  function getMuscleGroups() {
    return (
      allExercises &&
      allExercises
        .reduce((allGroups, exercise) => {
          const newGroups = [];
          const groupNames = exercise.muscle_groups.map(
            (muscle_group) => muscle_group.name
          );
          for (const group of groupNames) {
            !allGroups.includes(group) && newGroups.push(group);
          }
          return [...allGroups, ...newGroups];
        }, [])
        .sort()
    );
  }

  // Get exercises categorized by muscle group
  function getCategorizedExercises() {
    if (!allExercises) return allExercises;

    let categories = {};
    for (const muscleGroup of muscleGroups) {
      const exercises = allExercises
        .filter((exercise) => {
          const exerciseGroups = exercise.muscle_groups.map(
            (group) => group.name
          );
          return exerciseGroups.includes(muscleGroup);
        })
        .sort((a, b) => {
          const nameA = a.name.toUpperCase();
          const nameB = b.name.toUpperCase();
          if (nameA > nameB) return 1;
          if (nameA < nameB) return -1;
          return 0;
        });
      categories[muscleGroup] = exercises;
    }
    return categories;
  }

  return (
    <ExercisesContext.Provider
      value={{
        allExercises,
        muscleGroups,
        categorizedExercises,
        setAllExercises,
      }}
      {...props}
    />
  );
}

const useExercises = () => useContext(ExercisesContext);
export { ExercisesProvider, useExercises };
