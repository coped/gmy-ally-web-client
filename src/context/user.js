import { createContext, useContext } from "react";

const UserContext = createContext();
const useUser = () => useContext(UserContext);

export { UserContext, useUser };
