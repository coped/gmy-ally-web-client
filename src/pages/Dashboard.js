import React from "react";

export default function Dashboard() {
  return <p>This is the dashboard.</p>;
}
