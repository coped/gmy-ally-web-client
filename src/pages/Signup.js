import React, { useState } from "react";
import { Notification } from "components/common";
import { FormInput, FormButton } from "components/form";
import { Link, Redirect } from "react-router-dom";
import Api from "lib/api";
import { useAuth } from "context/auth";
import "assets/Signup.scss";

export default function Signup(props) {
  const [isLoading, setIsLoading] = useState(false);
  const [form, setForm] = useState({
    name: "",
    email: "",
    password: "",
    password_confirmation: "",
  });
  const [apiMessages, setApiMessages] = useState("");
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const { setAuthToken } = useAuth();

  async function signupUser(e) {
    e.preventDefault();
    setIsLoading(true);
    const data = await Api.createUser(form);
    if (data.status === "success") {
      setAuthToken(data.payload.jwt);
      setIsLoggedIn(true);
    } else {
      setApiMessages(data.messages);
      setIsLoading(false);
    }
  }

  function onChange(event) {
    const target = event.target;
    setForm({ ...form, [target.name]: target.value });
  }

  if (isLoggedIn) {
    return <Redirect to="/dashboard" />;
  }

  return (
    <div id="Signup">
      <h1 className="signup-title">Sign up</h1>
      <form onSubmit={signupUser} className="column is-6 box">
        {apiMessages &&
          apiMessages.map((message, index) => (
            <Notification key={index} type={message.type}>
              <p>{message.message}</p>
            </Notification>
          ))}
        <FormInput
          id="name-input"
          label="Name:"
          name="name"
          type="text"
          placeholder="Your name"
          value={form.name}
          onChange={onChange}
        />
        <FormInput
          id="email-input"
          label="Email:"
          name="email"
          type="email"
          placeholder="your@email.com"
          value={form.email}
          onChange={onChange}
        />
        <FormInput
          id="password-input"
          label="Password:"
          name="password"
          type="password"
          value={form.password}
          onChange={onChange}
        />
        <FormInput
          id="password-confirmation-input"
          label="Password confirmation:"
          name="passwordConfirmation"
          type="password"
          value={form.passwordConfirmation}
          onChange={onChange}
        />
        <FormButton classList={["is-link"]} loading={isLoading}>
          Sign up
        </FormButton>
        <p className="is-size-6">
          Already have an account? <Link to="/login">Log in</Link>.
        </p>
      </form>
    </div>
  );
}
