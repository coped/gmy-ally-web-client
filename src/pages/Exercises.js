import React, { useEffect, useState } from "react";
import { Api } from "lib/api";
import { PageLoader } from "components/common";
import { useExercises } from "context/exercises";
import "bulma-quickview/dist/css/bulma-quickview.min.css";
import bulmaQuickview from "bulma-quickview/dist/js/bulma-quickview.min.js";

export default function Exercises() {
  const {
    allExercises,
    muscleGroups,
    categorizedExercises,
    setAllExercises,
  } = useExercises();
  const [loading, setLoading] = useState(false);
  const [currentMuscleGroup, setCurrentMuscleGroup] = useState("abdominals");
  const [quickviewExercise, setQuickviewExercise] = useState();

  // Fetch exercise data on component mount
  useEffect(() => {
    async function fetchData() {
      setLoading(true);
      const data = await Api.indexExercises();
      setAllExercises(data.payload.exercises);
      setQuickviewExercise(data.payload.exercises[0]);
      setLoading(false);
    }
    !allExercises && fetchData();
  }, [allExercises, setAllExercises]);

  useEffect(() => {
    if (allExercises && !quickviewExercise)
      setQuickviewExercise(allExercises[0]);
    if (quickviewExercise) bulmaQuickview.attach();
  });

  function showExercises() {
    console.log(allExercises);
    allExercises && console.log(categorizedExercises);
  }

  function capitalize(string) {
    return string[0].toUpperCase() + string.substring(1);
  }

  return (
    <div id="Exercises">
      <PageLoader loading={loading} className="is-info" />
      {quickviewExercise && (
        <div id="quickviewDefault" className="quickview">
          <header className="quickview-header">
            <p className="title">{quickviewExercise.title}</p>
            <span className="delete" data-dismiss="quickview"></span>
          </header>

          <div className="quickview-body">
            <div className="quickview-block section">
              <h3 className="is-size-5">Movement type:</h3>
              <p>{quickviewExercise.movement_type}</p>
              <h3 className="is-size-5">Equipment:</h3>
              <div className="content">
                <ul>
                  {quickviewExercise.equipment.split("\n").map((item) => (
                    <li key={item}>{capitalize(item)}</li>
                  ))}
                </ul>
              </div>
              <h3 className="is-size-5">Steps:</h3>
              <div className="content">
                <ol>
                  {quickviewExercise.steps.split("\n").map((step, index) => (
                    <li key={"step-" + index}>{step}</li>
                  ))}
                </ol>
              </div>
            </div>
          </div>

          <footer className="quickview-footer">GYM PARTNER</footer>
        </div>
      )}
      <div className="columns">
        <div className="column is-3">
          <aside className="menu">
            <p className="menu-label">SEARCH EXERCISES BY MUSCLE GROUP</p>
            <ul className="menu-list">
              {muscleGroups &&
                muscleGroups.map((name) => (
                  <li key={name}>
                    <a onClick={() => setCurrentMuscleGroup(name)}>
                      {capitalize(name)}
                    </a>
                  </li>
                ))}
            </ul>
          </aside>
        </div>
        <div className="column">
          <h3 className="is-size-3 has-text-centered">
            {capitalize(currentMuscleGroup)}
          </h3>
          <div className="box">
            <ul className="menu-list">
              {categorizedExercises &&
                categorizedExercises[currentMuscleGroup].map((exercise) => (
                  <li className="list-item" key={exercise.name}>
                    <a
                      onClick={() => setQuickviewExercise(exercise)}
                      data-show="quickview"
                      data-target="quickviewDefault"
                    >
                      {exercise.title}
                    </a>
                  </li>
                ))}
            </ul>
          </div>
        </div>
      </div>
      <button onClick={showExercises} className="button">
        Show exercises
      </button>
    </div>
  );
}
