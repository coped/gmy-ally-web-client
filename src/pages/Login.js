import React, { useState } from "react";
import { Link, Redirect } from "react-router-dom";
import { FormInput, FormButton } from "components/form";
import { Notification } from "components/common";
import { useAuth } from "context/auth";
import { useUser } from "context/user";
import Api from "lib/api";
import "assets/Login.scss";

export default function Login() {
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [apiMessages, setApiMessages] = useState([]);
  const [form, setForm] = useState({ email: "", password: "" });

  const { setAuthContext } = useAuth();
  const { setUserContext } = useUser();

  function onChange(event) {
    const target = event.target;
    setForm({
      ...form,
      [target.name]: target.value,
    });
  }

  async function loginUser(e) {
    e.preventDefault();
    setIsLoading(true);
    const data = await Api.login({ info: form });
    if (data.status === "success") {
      setAuthContext(data.payload.jwt);
      setUserContext(data.payload.user);
      setIsLoggedIn(true);
    } else {
      setApiMessages(data.messages);
      setIsLoading(false);
    }
  }

  if (isLoggedIn) {
    return <Redirect to="/dashboard" />;
  }

  return (
    <div id="Login">
      <h1 className="login-title">Log in</h1>
      <form onSubmit={loginUser} className="column is-6 box">
        {apiMessages &&
          apiMessages.map((response, index) => (
            <Notification key={index} classList={[`is-${response.type}`]}>
              <p>{response.message}</p>
            </Notification>
          ))}
        <FormInput
          id="email-input"
          label="Email:"
          name="email"
          type="email"
          placeholder="your@email.com"
          value={form.email}
          onChange={onChange}
        />
        <FormInput
          id="password-input"
          label="Password:"
          name="password"
          type="password"
          value={form.password}
          onChange={onChange}
        />
        <FormButton classList={["is-link"]} loading={isLoading}>
          Log in
        </FormButton>
        <p>
          Don't have an account? <Link to="/signup">Sign up</Link>.
        </p>
      </form>
    </div>
  );
}
