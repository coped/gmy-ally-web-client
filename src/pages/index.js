import Login from "pages/Login";
import Signup from "pages/Signup";
import Welcome from "pages/Welcome";
import Dashboard from "pages/Dashboard";
import Exercises from "pages/Exercises";

export { Login, Signup, Welcome, Dashboard, Exercises };
