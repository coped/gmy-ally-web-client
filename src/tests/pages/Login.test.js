import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Enzyme, { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Login } from "pages";
import { AuthContext } from "context/auth";

Enzyme.configure({ adapter: new Adapter() });

describe("Login", () => {
  const withContext = (component) => (
    <AuthContext.Provider
      value={{ authTokens: "token", setAuthTokens: () => true }}
    >
      {component}
    </AuthContext.Provider>
  );

  it("renders a login form", () => {
    const wrapper = mount(withContext(<Login />));
    expect(wrapper.find("form").length).toEqual(1);
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(withContext(<Login />), div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("changes to input are reflected in login state", () => {
    const wrapper = mount(withContext(<Login />));
    const changeEmail = { target: { name: "email", value: "my-email" } };
    const changePassword = {
      target: { name: "password", value: "my-password" },
    };
    wrapper.find("input[type='email']").simulate("change", changeEmail);
    wrapper
      .find("input[type='password-input']")
      .simulate("change", changePassword);
    expect(wrapper.state().form.email).toEqual("my-email");
    expect(wrapper.state().form.password).toEqual("my-password");
  });

  it("has a valid snapshot", () => {
    const component = renderer.create(withContext(<Login />));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
