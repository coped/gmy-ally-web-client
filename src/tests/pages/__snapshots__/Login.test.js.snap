// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login has a valid snapshot 1`] = `
<div
  id="Login"
>
  <h1
    className="login-title"
  >
    Log in
  </h1>
  <form
    className="column is-6 box"
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <label
        className="label"
      >
        Email:
      </label>
      <div
        className="control"
      >
        <input
          className="input"
          id="email-input"
          name="email"
          onChange={[Function]}
          placeholder="your@email.com"
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Password:
      </label>
      <div
        className="control"
      >
        <input
          className="input"
          id="password-input"
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
    >
      <div
        className="control"
      >
        <button
          className="button is-link"
        >
          Log in
        </button>
      </div>
    </div>
    <p>
      Don't have an account? 
      <a
        href="/signup"
        onClick={[Function]}
      >
        Sign up
      </a>
      .
    </p>
  </form>
</div>
`;
